handler.go

```diff
@@ -71,44 +64,34 @@ func (redriver Redriver) HandleMessages(messages []events.SQSMessage, processor
 		sqsConnector = sqs.New(awsSession)
 	}
 
-	messagesCount := len(messages)
-	var processedMessages []processResult
-	var failures []processResult
-
-	processResultChannel := make(chan processResult)
+	processedMessagesResult := newProcessResults(messages)
+	processResultChannel := make(chan *processResult)
 	defer close(processResultChannel)
 
 	for _, message := range messages {
 		redriver.processMessageAsync(message, processor, processResultChannel)
 	}
 
-	for i := 0; i < messagesCount; i++ {
-		processResult := <-processResultChannel
-		if processResult.err != nil {
-			failures = append(failures, processResult)
-			continue
-		}
-		processedMessages = append(processedMessages, processResult)
+	for i := 0; i < processedMessagesResult.messagesCount; i++ {
+		res := <-processResultChannel
+		processedMessagesResult.addResult(res)
 	}
 
-	// All messages processed.
-	if len(processedMessages) == messagesCount {
+	if processedMessagesResult.hasOnlyProcessedMessages() {
 		return nil
 	}
 
-	// All messages failed.
-	if len(failures) == messagesCount {
-		return fmt.Errorf("all messages processing failed, %+v", failures)
+	if processedMessagesResult.hasProcessFailedMessages() {
```

Comment:
This should only trigger if there is only failed messages not if there is some failed messages, that is handled line 96

Reply:
Why would it matter to know if there was at least 1 error vs only errors? Is there any business value knowing that we had to redrive 1 or all the messages?

Reply:
Because if you return now and have a mix of failed and successful you do not delete the successfully processed messages from the queue (see line 92)

Reply:
`if processedMessagesResult.hasOnlySuccessfulMessages() {
		return nil
	}

	if err := redriver.deleteProcessedMessages(*processedMessagesResult, sqsConnector); !redriver.Debug && err != nil {
		return err
	}

	return processResultsError{Results: *processedMessagesResult}`

Does it look ok to you? I was thinking that when there are errors, there is no need to check whether there are only errors or not. We assume that we always need to delete successful messages. If there is no successful message, we will loop through an empty slice which is not ideal but it makes the code a little bit more straight forward.

Reply:
There is no need to delete messages when they all failed you just need to make sure the lambda is returning an error. Better performance

Reply:
Oh yes ok then !